crates/core/src/server_session.rs

```diff
@@ -0,0 +1,243 @@
+use anyhow::{anyhow, bail, Result};
+use chrono::{DateTime, Utc};
+use governor::clock::DefaultClock;
+use governor::state::{InMemoryState, NotKeyed};
+use governor::RateLimiter;
+use rand::Rng;
+use std::collections::VecDeque;
+use std::convert::TryFrom;
+use std::fmt;
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex, RwLock};
+
+use crate::identity::Identity;
+use ya_client_model::NodeId;
+use ya_relay_proto::proto;
+use ya_relay_proto::proto::{SlotId, SESSION_ID_SIZE};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, derive_more::Display)]
+pub enum TransportType {
+    Unreliable,
+    Reliable,
+    Transfer,
+}
+
+#[derive(Copy, Clone, PartialEq, PartialOrd, Hash, Eq, Ord)]
+pub struct SessionId {
+    id: [u8; SESSION_ID_SIZE],
+}
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct Endpoint {
+    pub protocol: proto::Protocol,
+    pub address: SocketAddr,
+}
+
+#[derive(Clone)]
+pub struct NodeInfo {
+    pub identities: Vec<Identity>,
+    pub slot: SlotId,
+
+    /// Endpoints registered by Node.
+    pub endpoints: Vec<Endpoint>,
+    pub supported_encryption: Vec<String>,
+}
+
+impl NodeInfo {
+    pub fn node_id(&self) -> NodeId {
+        self.identities.get(0).map(|ident| ident.node_id).unwrap()
```

Comment:
1. I think it would be better to utilize rust type system to have default identity and then secondary as a vector separately
2. this method probably should be renamed to `default_node_id` as far as I understand